### PR TITLE
Add OVAL variable references to yaml rules. [WIP]

### PR DESCRIPTION
**Pull request by matejak**
During the yaml input implementation, the OVAL variable references were omitted, as @dahaic noticed.
This PR extends the yaml rule format and adds variable references back, so the functionality is preserved.

### CODE REVIEW

File: shared/utils/yaml-to-shorthand.py [hunk 1]
```diff
@@ -422,6 +423,14 @@ def to_xml_element(self):
         oval_ref = ET.SubElement(rule, "oval")
         oval_ref.set("id", self.id_)
 
+        oval_vars_count = len(self.oval_variables)
+        if oval_vars_count > 0:
+            if oval_vars_count > 1:
```
[the remark this answers is not in the document]
Comment (matejak) [reply shared/utils/yaml-to-shorthand.py]: Me and @dahaic weren't sure whether there can be more than one referenced OVAL variables per one rule. This code makes sure that an error is at least printed when one tries to use more than one variable per rule, which isn't supported.

**Comment by mpreisler** [comment]
I don't like that we have to list which variables to put into OVAL, I think that's info that we already have. Any reason why the script can't just figure out what OVAL wants and add the check-exports? This is something that will make development harder and yet another opportunity to misspell IDs.

**Comment by matejak** [comment]
I don't know what is the optimal procedure. However, I am sure that this is a relatively easy and fast way of fixing the SSG master, which is broken due to missing variable references. I also think that this way of fixing stuff is not fundamentally wrong and as a better way of fixing this issue is developed, rolling back changes should be easy enough.

**Comment by matejak** [comment]
BTW I don't see the `guide` directory under `rhel7`, is this OK, or something is missing? I am trying to test `rhel7` profile with the test suite and there are lots of rules having the `other` status (which indicates that an OVAL variable is missing).

**Comment by mpreisler** [comment]
> BTW I don't see the guide directory under rhel7, is this OK, or something is missing? I am trying to test rhel7 profile with the test suite and there are lots of rules having the other status (which indicates that an OVAL variable is missing).

It's using shared/guide. see product.yml

**Comment by mpreisler** [comment]
> I don't know what is the optimal procedure. However, I am sure that this is a relatively easy and fast way of fixing the SSG master, which is broken due to missing variable references. I also think that this way of fixing stuff is not fundamentally wrong and as a better way of fixing this issue is developed, rolling back changes should be easy enough.

Please let's go for optimal, it should be easy to do the right thing and avoid adding error-prone boilerplate to yaml. We know which variables OVAL needs, they are listed in the OVAL definition. We can just parse that and fill in check-exports. We had to list this before but it is clearly better to automate what can be automated and not require busy work from devs.

File: shared/utils/yaml-to-shorthand.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (mpreisler) [reply shared/utils/yaml-to-shorthand.py]: You can export any number of XCCDF Values as OVAL variables.

**Comment by dahaic** [comment]
@mpreisler I disagree - optimum is very volatile target, and in this case the proposed optimization makes the `yaml-to-shorthand` much more "clever". As we need to accept bunch of assumptions and limitations.

* oval is already created (thus we run yaml-to-shorthand after templating)
* oval is created in the target (as shared rule can have product based oval)
* introduction of xml parsing into the "yaml->shorthand" script, which makes the logic actually quite convoluted in my opinion

And all that while master is completely broken. With that, I would prefer elegant hotfix, and then we can search for optimums that won't assume too much, without time pressure.

**Comment by mpreisler** [comment]
> And all that while master is completely broken. With that, I would prefer elegant hotfix, and then we can search for optimums that won't assume too much, without time pressure.

Yes, elegant hotfix would be great. I don't think changing 550 yamls is elegant though :) And I fear that if this gets in it never gets out.

Remember, we don't have to do this at yaml phase. In fact I think that would be an inferior solution. This should be done at linking time IMO, that's when all the files are parsed and you can do these connections fairly easily with little performance impact.

**Comment by matejak** [comment]
I disagree about the PR - it is a patch which fixes something and which can be taken out later easily, as it modifies files in a well-defined way.
Anyway, I would love to implement a better solution, but I don't know anything about the part of SSG build system that should be modifie. As it is not obvious to me what is the purpose behind the `relabel-ids.py` file is and the code has relatively high cyclomatic complexity, I feel that I should get some introduction and background info before going into the implementation.

File: shared/utils/yaml-to-shorthand.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (yuumasato) [reply shared/utils/yaml-to-shorthand.py]: We have very few rules that export more than one OVAL variable, as an example there is Rule `sshd_set_timeout`, https://github.com/OpenSCAP/scap-security-guide/blob/7334ef8e3e0b452843e912c5f1b69d361f8c7bfb/rhel6/xccdf/services/ssh.xml#L209 

First variable is consumed by OVAL definition referenced by Rule[1], and second variable is consumed by the extended definitions, for example [2].

[1] https://github.com/OpenSCAP/scap-security-guide/blob/master/shared/checks/oval/sshd_set_idle_timeout.xml

[2] https://github.com/OpenSCAP/scap-security-guide/blob/master/shared/checks/oval/sshd_not_required_or_unset.xml

**Comment by matejak** [comment]
The PR #2704 makes it feasible to add the OVAL variables autofill into XCCDF to `relabel-ids.py`. However, this requires the ability to resolve an OVAL definition, so one can discover variable references that may be buried quite deep (definition references test that references object that references the variable). Do we have such functionality, or do we have to develop it?

**Comment by redhatrises** [comment]
This PR is really needed as the datastreams are erroring out all over the place.

**Comment by matejak** [comment]
I am now working on a proper solution, so maybe we can survive a couple more days until it is fixed for good.

**Pull request closed without merging**